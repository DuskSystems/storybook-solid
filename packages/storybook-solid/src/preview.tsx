// biome-ignore-all lint/suspicious/noExplicitAny: Copied from React integration.
// biome-ignore-all lint/complexity/noBannedTypes: Copied from React integration.

import type { Component as ComponentType } from "solid-js";
import type { Meta, Preview, Story } from "storybook/internal/csf";
import type {
  Args,
  ArgsStoryFn,
  ComponentAnnotations,
  DecoratorFunction,
  Renderer,
  StoryAnnotations,
} from "storybook/internal/types";
import type { OmitIndexSignature, SetOptional, Simplify, UnionToIntersection } from "type-fest";
import { __definePreview as definePreviewBase } from "./csf4";
import * as solidAnnotations from "./entry-preview";
import * as solidDocsAnnotations from "./entry-preview-docs";
import type { AddMocks } from "./public-types";
import type { SolidRenderer } from "./types";

export function __definePreview(preview: SolidPreview["input"]) {
  return definePreviewBase({
    ...preview,
    // @ts-expect-error: Copied from React integration.
    addons: [solidAnnotations, solidDocsAnnotations, ...(preview.addons ?? [])],
  }) as SolidPreview;
}

export interface SolidPreview extends Preview<SolidRenderer> {
  meta<TArgs extends Args, Decorators extends DecoratorFunction<SolidRenderer, any>, TMetaArgs extends Partial<TArgs>>(
    meta: {
      render?: ArgsStoryFn<SolidRenderer, TArgs>;
      component?: ComponentType<TArgs>;
      decorators?: Decorators | Decorators[];
      args?: TMetaArgs;
    } & Omit<ComponentAnnotations<SolidRenderer, TArgs>, "decorators">,
  ): SolidMeta<
    {
      args: Simplify<TArgs & Simplify<OmitIndexSignature<DecoratorsArgs<SolidRenderer, Decorators>>>>;
    },
    { args: Partial<TArgs> extends TMetaArgs ? {} : TMetaArgs }
  >;
}

type DecoratorsArgs<TRenderer extends Renderer, Decorators> = UnionToIntersection<
  Decorators extends DecoratorFunction<TRenderer, infer TArgs> ? TArgs : unknown
>;

interface SolidMeta<Context extends { args: Args }, MetaInput extends ComponentAnnotations<SolidRenderer>>
  extends Meta<SolidRenderer, Context["args"]> {
  story<
    TInput extends StoryAnnotations<SolidRenderer, Context["args"]> & {
      render: () => SolidRenderer["storyResult"];
    },
  >(story: TInput): SolidStory;

  story<
    TInput extends Simplify<
      StoryAnnotations<
        SolidRenderer,
        AddMocks<Context["args"], MetaInput["args"]>,
        SetOptional<Context["args"], keyof Context["args"] & keyof MetaInput["args"]>
      >
    >,
  >(story: TInput): SolidStory;
}

export interface SolidStory extends Story<SolidRenderer> {}
