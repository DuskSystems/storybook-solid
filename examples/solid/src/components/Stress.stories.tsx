import { Show } from "solid-js";
import preview from "#.storybook/preview";
import { Stress } from "./Stress";

const meta = preview.meta({
  id: "stress-component",
  title: "Example/Stress",
  component: Stress,
  tags: ["autodocs"],
});

export const Example = meta.story({
  args: {
    string: "hello",
    number: 42,
    boolTrue: true,
    boolFalse: false,
    nil: null,
    array: [1, 2, 3],
    nested: ["nested", ["array", 456]],
    object: { foo: "bar", baz: false },
    complex: { bar: 123, foo: ["array", 456] },
    func: function power(x: number) {
      return x * x;
    },
    arrow: (x: number) => x * 2,
    element: <div>Hi</div>,
    symbol: Symbol("foo"),
    children: [
      "text",
      123,
      true,
      false,
      null,
      ["nested", ["array", 456]],
      // { foo: "bar" },
      <div>Hello</div>,
      <span>World</span>,
      // () => <em>JSX from function</em>,
      // () => 99,
      <Show when={() => true}>Show</Show>,
    ],
  },
});
