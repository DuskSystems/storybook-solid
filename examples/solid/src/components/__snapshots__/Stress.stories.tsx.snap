// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`All Attributes > AllAttributes 1`] = `
<div>
  <div>
    <div>
      Hi
    </div>
  </div>
  <pre>
    {"string":"hello","number":42,"boolTrue":true,"boolFalse":false,"nil":null,"array":[1,2,3],"nested":["nested",["array",456]],"object":{"foo":"bar","baz":false},"complex":{"bar":123,"foo":["array",456]}}
  </pre>
</div>
`;
